crates/db_schema/src/impls/actor_language.rs

```diff
@@ -0,0 +1,362 @@
+use crate::{
+  newtypes::{CommunityId, LanguageId, LocalUserId, SiteId},
+  source::{actor_language::*, community::Community, language::Language},
+};
+use diesel::{
+  delete,
+  dsl::*,
+  insert_into,
+  result::Error,
+  select,
+  ExpressionMethods,
+  PgConnection,
+  QueryDsl,
+  RunQueryDsl,
+};
+use lemmy_utils::error::LemmyError;
+
+impl LocalUserLanguage {
+  pub fn read(
+    conn: &mut PgConnection,
+    for_local_user_id: LocalUserId,
+  ) -> Result<Vec<LanguageId>, Error> {
+    use crate::schema::local_user_language::dsl::*;
+
+    local_user_language
+      .filter(local_user_id.eq(for_local_user_id))
+      .select(language_id)
+      .get_results(conn)
+  }
+
+  /// Update the user's languages.
+  ///
+  /// If no language_id vector is given, it will show all languages
+  pub fn update(
+    conn: &mut PgConnection,
+    language_ids: Vec<LanguageId>,
+    for_local_user_id: LocalUserId,
+  ) -> Result<(), Error> {
+    conn.build_transaction().read_write().run(|conn| {
+      use crate::schema::local_user_language::dsl::*;
+      // Clear the current user languages
+      delete(local_user_language.filter(local_user_id.eq(for_local_user_id))).execute(conn)?;
+
+      let lang_ids = update_languages(conn, language_ids)?;
+      for l in lang_ids {
+        let form = LocalUserLanguageForm {
+          local_user_id: for_local_user_id,
+          language_id: l,
+        };
+        insert_into(local_user_language)
+          .values(form)
+          .get_result::<Self>(conn)?;
+      }
+      Ok(())
+    })
+  }
+}
+
+impl SiteLanguage {
+  pub fn read_local(conn: &mut PgConnection) -> Result<Vec<LanguageId>, Error> {
+    use crate::schema::{site, site_language::dsl::*};
+    let subquery = crate::schema::site::dsl::site
+      .order_by(site::id)
+      .select(site::id)
+      .limit(1)
+      .into_boxed();
+    site_language
+      .filter(site_id.eq_any(subquery))
+      .select(language_id)
+      .load(conn)
+  }
+
+  pub fn read(conn: &mut PgConnection, for_site_id: SiteId) -> Result<Vec<LanguageId>, Error> {
+    use crate::schema::site_language::dsl::*;
+    site_language
+      .filter(site_id.eq(for_site_id))
+      .select(language_id)
+      .load(conn)
+  }
+
+  pub fn update(
+    conn: &mut PgConnection,
+    language_ids: Vec<LanguageId>,
+    for_site_id: SiteId,
+  ) -> Result<(), Error> {
+    conn.build_transaction().read_write().run(|conn| {
```

Comment:
With this transaction in place, many tests fail with error `Cannot perform this operation while a transaction is open`. But if i comment it out, tests pass just fine. Very strange.

Reply:
Its trying to do tests at the same time then, weird. I thought that `#serial` marker was supposed to mitigate that.

Reply:
Maybe some tests are missing the marker.

Reply:
There was a `#[serial]` missing in impls::person::tests::test_crud, but that didnt fix it. I suspect its because some apub tests use `#[actix_rt::test]` so its a different test runner and might run in parallel. Maybe we'll have to change all tests to use that annotiation. For now im leaving a todo here.

Reply:
Tried to fix it using block_on as in the link below, but didnt work.

https://stackoverflow.com/a/64597248

Reply:
Found it, this calls `CommunityLanguage::limit_languages()` which also has a transaction, so there were two nested transactions. Removed the inner one now and it works.